File: linux/fieldcommand/nav.py
"""Navigation grid and A* pathfinding (pure Python, no pygame).

The map is divided into 40-unit cells. A cell is blocked if its centre lies inside a wall, building, tree or
crystal, padded by CLEARANCE so unit bodies keep off the edges. Units walk straight at their goal when the line
is clear, and otherwise follow a smoothed A* path. The grid is rebuilt when buildings appear or disappear.
"""
import heapq
import math

from . import defs

CELL = 40.0
CLEARANCE = 14.0
_SQRT2 = math.sqrt(2)
_NEIGHBOURS = [(1, 0, 1.0), (-1, 0, 1.0), (0, 1, 1.0), (0, -1, 1.0),
               (1, 1, _SQRT2), (1, -1, _SQRT2), (-1, 1, _SQRT2), (-1, -1, _SQRT2)]


class NavGrid:
    def __init__(self):
        self.cols = int(math.ceil(defs.WORLD_W / CELL))
        self.rows = int(math.ceil(defs.WORLD_H / CELL))
        self.blocked = bytearray(self.cols * self.rows)
        self.version = 0

    # ------------------------------------------------------------ building the grid

    def rebuild(self, rects, circles):
        """rects: [(x0, y0, x1, y1)], circles: [(x, y, r)] of impassable space."""
        b = bytearray(self.cols * self.rows)
        cols, rows = self.cols, self.rows
        pad = CLEARANCE
        for (x0, y0, x1, y1) in rects:
            cx0 = max(0, int((x0 - pad) // CELL))
            cx1 = min(cols - 1, int((x1 + pad) // CELL))
            cy0 = max(0, int((y0 - pad) // CELL))
            cy1 = min(rows - 1, int((y1 + pad) // CELL))
            for cy in range(cy0, cy1 + 1):
                py = (cy + 0.5) * CELL
                if py < y0 - pad or py > y1 + pad:
                    continue
                row = cy * cols
                for cx in range(cx0, cx1 + 1):
                    px = (cx + 0.5) * CELL
                    if x0 - pad <= px <= x1 + pad:
                        b[row + cx] = 1
        for (x, y, r) in circles:
            rr = r + pad
            cx0, cx1 = max(0, int((x - rr) // CELL)), min(cols - 1, int((x + rr) // CELL))
            cy0, cy1 = max(0, int((y - rr) // CELL)), min(rows - 1, int((y + rr) // CELL))
            r2 = rr * rr
            for cy in range(cy0, cy1 + 1):
                dy = (cy + 0.5) * CELL - y
                for cx in range(cx0, cx1 + 1):
                    dx = (cx + 0.5) * CELL - x
                    if dx * dx + dy * dy <= r2:
                        b[cy * cols + cx] = 1
        self.blocked = b
        self.version += 1

    # ------------------------------------------------------------ queries

    def cell(self, x, y):
        return (min(self.cols - 1, max(0, int(x // CELL))), min(self.rows - 1, max(0, int(y // CELL))))

    def center(self, cx, cy):
        return ((cx + 0.5) * CELL, (cy + 0.5) * CELL)

    def is_blocked(self, cx, cy):
        return self.blocked[cy * self.cols + cx] == 1

    def line_clear(self, x0, y0, x1, y1, ignore_end=0.0):
        """True if the straight line crosses no blocked cell. The first cell and the final `ignore_end` units
        (e.g. the target building's own footprint) are not checked."""
        d = math.hypot(x1 - x0, y1 - y0)
        length = d - ignore_end
        if length <= 0:
            return True
        step = CELL * 0.4
        n = int(length / step)
        if n <= 0:
            return True
        ux, uy = (x1 - x0) / d, (y1 - y0) / d
        start = self.cell(x0, y0)
        cols, blocked = self.cols, self.blocked
        for i in range(1, n + 1):
            t = i * step
            cx, cy = int((x0 + ux * t) // CELL), int((y0 + uy * t) // CELL)
            if (cx, cy) == start:
                continue
            if cx < 0 or cy < 0 or cx >= cols or cy >= self.rows or blocked[cy * cols + cx]:
                return False
        return True

    def nearest_free(self, cx, cy, max_r=8):
        if not self.is_blocked(cx, cy):
            return cx, cy
        best, best_d = None, 1e9
        for r in range(1, max_r + 1):
            for dy in range(-r, r + 1):
                for dx in range(-r, r + 1):
                    if max(abs(dx), abs(dy)) != r:
                        continue
                    x, y = cx + dx, cy + dy
                    if 0 <= x < self.cols and 0 <= y < self.rows and not self.blocked[y * self.cols + x]:
                        d = dx * dx + dy * dy
                        if d < best_d:
                            best, best_d = (x, y), d
            if best:
                return best
        return cx, cy

    # ------------------------------------------------------------ A*

    def find_path(self, x0, y0, x1, y1, max_nodes=9000):
        """Waypoints from (x0, y0) toward (x1, y1), ending at the goal itself. If the goal is unreachable the
        path leads to the reachable cell closest to it. Returns [] when the line is already clear."""
        cols, rows, blocked = self.cols, self.rows, self.blocked
        sx, sy = self.nearest_free(*self.cell(x0, y0))
        gx, gy = self.nearest_free(*self.cell(x1, y1))
        start, goal = sy * cols + sx, gy * cols + gx
        if start == goal:
            return [(x1, y1)]

        def h(i):
            dx, dy = abs(i % cols - gx), abs(i // cols - gy)
            return (dx + dy) + (_SQRT2 - 2) * min(dx, dy)

        g = {start: 0.0}
        came = {}
        open_ = [(h(start), start)]
        closed = set()
        best, best_h = start, h(start)
        expanded = 0
        while open_:
            _, cur = heapq.heappop(open_)
            if cur in closed:
                continue
            if cur == goal:
                best = goal
                break
            closed.add(cur)
            expanded += 1
            if expanded > max_nodes:
                break
            hc = h(cur)
            if hc < best_h:
                best, best_h = cur, hc
            cx, cy = cur % cols, cur // cols
            gc = g[cur]
            for dx, dy, cost in _NEIGHBOURS:
                nx, ny = cx + dx, cy + dy
                if nx < 0 or ny < 0 or nx >= cols or ny >= rows:
                    continue
                ni = ny * cols + nx
                if blocked[ni] or ni in closed:
                    continue
                if dx and dy and (blocked[cy * cols + nx] or blocked[ny * cols + cx]):
                    continue  # no corner cutting
                ng = gc + cost
                if ng < g.get(ni, 1e18):
                    g[ni] = ng
                    came[ni] = cur
                    heapq.heappush(open_, (ng + h(ni), ni))
        # Reconstruct
        cells = [best]
        while cells[-1] in came:
            cells.append(came[cells[-1]])
        cells.reverse()
        pts = [self.center(i % cols, i // cols) for i in cells[1:]]
        if best == goal:
            pts.append((x1, y1))
        return self._smooth((x0, y0), pts)

    def _smooth(self, origin, pts):
        """String-pulling: skip waypoints that can be seen directly from the previous kept point."""
        if len(pts) <= 1:
            return pts
        out = []
        anchor = origin
        i = 0
        while i < len(pts):
            j = len(pts) - 1
            while j > i and not self.line_clear(anchor[0], anchor[1], pts[j][0], pts[j][1]):
                j -= 1
            out.append(pts[j])
            anchor = pts[j]
            i = j + 1
        return out

File: macos/Sources/FieldCommand/Nav.swift
import Foundation

/// Navigation grid and A* pathfinding for the server simulation; a port of the Linux edition's nav.py.
///
/// The map is divided into 40-unit cells. A cell is blocked if its centre lies inside a wall, building, tree or
/// crystal, padded by `clearance` so unit bodies keep off the edges. Units walk straight at their goal when the
/// line is clear, and otherwise follow a smoothed A* path. The grid is rebuilt when buildings appear or disappear.
final class SNavGrid {
    static let cell = 40.0
    static let clearance = 14.0
    private static let sqrt2 = 2.0.squareRoot()
    private static let neighbours: [(Int, Int, Double)] = [(1, 0, 1), (-1, 0, 1), (0, 1, 1), (0, -1, 1),
                                                           (1, 1, sqrt2), (1, -1, sqrt2), (-1, 1, sqrt2), (-1, -1, sqrt2)]
    let cols: Int, rows: Int
    private(set) var blocked: [Bool]
    private(set) var version = 0

    init() {
        cols = Int((worldW / Self.cell).rounded(.up))
        rows = Int((worldH / Self.cell).rounded(.up))
        blocked = Array(repeating: false, count: cols * rows)
    }

    // MARK: Building the grid

    func rebuild(rects: [SRect], circles: [(Double, Double, Double)]) {
        var b = [Bool](repeating: false, count: cols * rows)
        let pad = Self.clearance, cell = Self.cell
        for r in rects {
            let cx0 = max(0, Int(((r.x0 - pad) / cell).rounded(.down))), cx1 = min(cols - 1, Int((r.x1 + pad) / cell))
            let cy0 = max(0, Int(((r.y0 - pad) / cell).rounded(.down))), cy1 = min(rows - 1, Int((r.y1 + pad) / cell))
            guard cx0 <= cx1, cy0 <= cy1 else { continue }
            for cy in cy0...cy1 {
                let py = (Double(cy) + 0.5) * cell
                if py < r.y0 - pad || py > r.y1 + pad { continue }
                for cx in cx0...cx1 {
                    let px = (Double(cx) + 0.5) * cell
                    if px >= r.x0 - pad && px <= r.x1 + pad { b[cy * cols + cx] = true }
                }
            }
        }
        for (x, y, r) in circles {
            let rr = r + pad, r2 = rr * rr
            let cx0 = max(0, Int(((x - rr) / cell).rounded(.down))), cx1 = min(cols - 1, Int((x + rr) / cell))
            let cy0 = max(0, Int(((y - rr) / cell).rounded(.down))), cy1 = min(rows - 1, Int((y + rr) / cell))
            guard cx0 <= cx1, cy0 <= cy1 else { continue }
            for cy in cy0...cy1 {
                let dy = (Double(cy) + 0.5) * cell - y
                for cx in cx0...cx1 {
                    let dx = (Double(cx) + 0.5) * cell - x
                    if dx * dx + dy * dy <= r2 { b[cy * cols + cx] = true }
                }
            }
        }
        blocked = b
        version += 1
    }

    // MARK: Queries

    func cell(_ x: Double, _ y: Double) -> (Int, Int) {
        (min(cols - 1, max(0, Int((x / Self.cell).rounded(.down)))), min(rows - 1, max(0, Int((y / Self.cell).rounded(.down)))))
    }

    func center(_ cx: Int, _ cy: Int) -> (Double, Double) { ((Double(cx) + 0.5) * Self.cell, (Double(cy) + 0.5) * Self.cell) }

    /// True if the straight line crosses no blocked cell. The first cell and the final `ignoreEnd` units
    /// (e.g. the target building's own footprint) are not checked.
    func lineClear(_ x0: Double, _ y0: Double, _ x1: Double, _ y1: Double, ignoreEnd: Double = 0) -> Bool {
        let d = ((x1 - x0) * (x1 - x0) + (y1 - y0) * (y1 - y0)).squareRoot()
        let length = d - ignoreEnd
        if length <= 0 { return true }
        let step = Self.cell * 0.4
        let n = Int(length / step)
        if n <= 0 { return true }
        let ux = (x1 - x0) / d, uy = (y1 - y0) / d
        let start = cell(x0, y0)
        for i in 1...n {
            let t = Double(i) * step
            let fx = ((x0 + ux * t) / Self.cell).rounded(.down), fy = ((y0 + uy * t) / Self.cell).rounded(.down)
            let cx = Int(fx), cy = Int(fy)
            if cx == start.0 && cy == start.1 { continue }
            if cx < 0 || cy < 0 || cx >= cols || cy >= rows || blocked[cy * cols + cx] { return false }
        }
        return true
    }

    func nearestFree(_ cx: Int, _ cy: Int, maxR: Int = 8) -> (Int, Int) {
        if !blocked[cy * cols + cx] { return (cx, cy) }
        for r in 1...maxR {
            var best: (Int, Int)?, bestD = Int.max
            for dy in -r...r {
                for dx in -r...r where max(abs(dx), abs(dy)) == r {
                    let x = cx + dx, y = cy + dy
                    if x >= 0 && y >= 0 && x < cols && y < rows && !blocked[y * cols + x] {
                        let d = dx * dx + dy * dy
                        if d < bestD { best = (x, y); bestD = d }
                    }
                }
            }
            if let best { return best }
        }
        return (cx, cy)
    }

    // MARK: A*

    /// Waypoints from (x0, y0) toward (x1, y1), ending at the goal itself. If the goal is unreachable the path
    /// leads to the reachable cell closest to it.
    func findPath(_ x0: Double, _ y0: Double, _ x1: Double, _ y1: Double, maxNodes: Int = 9000) -> [(Double, Double)] {
        let s0 = cell(x0, y0), g0 = cell(x1, y1)
        let (sx, sy) = nearestFree(s0.0, s0.1), (gx, gy) = nearestFree(g0.0, g0.1)
        let start = sy * cols + sx, goal = gy * cols + gx
        if start == goal { return [(x1, y1)] }
        let cols = self.cols, rows = self.rows
        func h(_ i: Int) -> Double {
            let dx = Double(abs(i % cols - gx)), dy = Double(abs(i / cols - gy))
            return dx + dy + (Self.sqrt2 - 2) * min(dx, dy)
        }
        var g = [Double](repeating: .infinity, count: cols * rows)
        var came = [Int](repeating: -1, count: cols * rows)
        var closed = [Bool](repeating: false, count: cols * rows)
        var heap = MinHeap()
        g[start] = 0
        heap.push(h(start), start)
        var best = start, bestH = h(start), expanded = 0
        while let (_, cur) = heap.pop() {
            if closed[cur] { continue }
            if cur == goal { best = goal; break }
            closed[cur] = true
            expanded += 1
            if expanded > maxNodes { break }
            let hc = h(cur)
            if hc < bestH { best = cur; bestH = hc }
            let cx = cur % cols, cy = cur / cols
            for (dx, dy, cost) in Self.neighbours {
                let nx = cx + dx, ny = cy + dy
                if nx < 0 || ny < 0 || nx >= cols || ny >= rows { continue }
                let ni = ny * cols + nx
                if blocked[ni] || closed[ni] { continue }
                if dx != 0 && dy != 0 && (blocked[cy * cols + nx] || blocked[ny * cols + cx]) { continue }  // no corner cutting
                let ng = g[cur] + cost
                if ng < g[ni] {
                    g[ni] = ng
                    came[ni] = cur
                    heap.push(ng + h(ni), ni)
                }
            }
        }
        var cells = [best]
        while came[cells[cells.count - 1]] >= 0 { cells.append(came[cells[cells.count - 1]]) }
        cells.reverse()
        var pts = cells.dropFirst().map { center($0 % cols, $0 / cols) }
        if best == goal { pts.append((x1, y1)) }
        return smooth((x0, y0), pts)
    }

    /// String-pulling: skip waypoints that can be seen directly from the previous kept point.
    private func smooth(_ origin: (Double, Double), _ pts: [(Double, Double)]) -> [(Double, Double)] {
        if pts.count <= 1 { return pts }
        var out: [(Double, Double)] = []
        var anchor = origin
        var i = 0
        while i < pts.count {
            var j = pts.count - 1
            while j > i && !lineClear(anchor.0, anchor.1, pts[j].0, pts[j].1) { j -= 1 }
            out.append(pts[j])
            anchor = pts[j]
            i = j + 1
        }
        return out
    }
}

/// Binary min-heap of (priority, cell index).
private struct MinHeap {
    private var items: [(Double, Int)] = []

    mutating func push(_ p: Double, _ v: Int) {
        items.append((p, v))
        var i = items.count - 1
        while i > 0 {
            let parent = (i - 1) / 2
            if items[parent].0 <= items[i].0 { break }
            items.swapAt(parent, i)
            i = parent
        }
    }

    mutating func pop() -> (Double, Int)? {
        guard !items.isEmpty else { return nil }
        let top = items[0]
        let last = items.removeLast()
        if !items.isEmpty {
            items[0] = last
            var i = 0
            while true {
                let l = 2 * i + 1, r = l + 1
                var m = i
                if l < items.count && items[l].0 < items[m].0 { m = l }
                if r < items.count && items[r].0 < items[m].0 { m = r }
                if m == i { break }
                items.swapAt(m, i)
                i = m
            }
        }
        return top
    }
}
